RD & route-target properties move to vrf

chrisvanheuveln opened the pull request:
- route-distinguisher is now supported in both vrf and bgp
- route-target props have moved completely into vrf_af (new provider)
- vrf.yaml now template-tized
- changed feature calls to use new Feature class
- minitests updated
- feature calls moved to feature.rb

chrisvanheuveln committed:
RD & route-target properties move to vrf

 * route-distinguisher is now supported in both vrf and bgp
 * route-target props have moved completely into vrf_af (new provider)
 * vrf.yaml now template-tized
 * changed feature calls to use new Feature class
 * minitests updated

chrisvanheuveln committed:
Merge pull request #168 from bansalpradeep/develop

changing default value of acl prop

glennmatthews commented on lib/cisco_node_utils/feature.rb:
```diff
@@ -19,24 +19,69 @@
+      config_get('feature', 'bgp')
     rescue Cisco::CliError => e
-      # cmd will syntax when feature is not enabled.
+      # cmd will syntax reject when feature is not enabled
```
This seems weird - config_get for bgp is just calling 'show run | i feature' so why would we get a CliError in this case?

chrisvanheuveln replied:
It's a cut&paste from the old code which was doing show run bgp. I'll remove it and the others where appropriate.

chrisvanheuveln commented on lib/cisco_node_utils/bgp.rb:
```diff
@@ -119,32 +98,15 @@ def router_bgp(asnum, vrf, state='')
-      else
-        enable_create_router_bgp(@asnum, @vrf)
-      end
+      Feature.bgp_enable unless Feature.bgp_enabled?
```
@glennmatthews Just had a thought about these  `unless _enabled?` checks. I'm thinking of just removing them from the callers and add the unless checks to the `_enable` methods themselves.  Is it worth changing?

edit: nm, changed it anyway.

chrisvanheuveln committed:
Bugfix for vlan_mapping getter

Symptom:
  N3k gets error: undefined method `each' for nil:NilClass

Problem:
  config_get should return the YAML default_value if it doesn't find a match
  but the platform exclude causes it to return nil instead.

  config_get('interface', 'vlan_mapping', @name).each(&:compact!)

  vlan_mapping:
    _exclude: [/N(3|5|6|9)/]
    default_value: []

Minitest didn't catch it because the test was skipped due to a platform check.

chrisvanheuveln committed:
Move feature calls into feature.rb

Remove support checks from minitests

chrisvanheuveln committed:
Merge branch 'mv-rd-rt-from-bgp-to-vrf' of https://github.com/chrisvanheuveln/cisco-network-node-utils into mv-rd-rt-from-bgp-to-vrf

Conflicts:
	tests/test_vxlan_vtep.rb

robert-w-gries commented on lib/cisco_node_utils/bgp.rb:
```diff
@@ -691,16 +653,15 @@ def default_reconnect_interval
-      RouterBgpAF.feature_nv_overlay_evpn_enable unless
-        RouterBgpAF.feature_nv_overlay_evpn_enabled
+      Feature.nv_overlay_enable
+      Feature.nv_overlay_evpn_enable
```
Nit: Since the `nv_overlay_evpn` feature depends on `nv_overlay` being enabled, you can move `nv_overlay_enable` to the `Feature.nv_overlay_evpn_enable` method

chrisvanheuveln replied:
Yeah, considered that. Also thought about maybe getting rid of nv_overlay_evpn_enable altogether and just making it part of nv_overlay_enable since it's not really a "feature" command (`nv overlay evpn`). I think I'd prefer to keep it like this until we can come up with a better story for it.

robert-w-gries commented on lib/cisco_node_utils/cmd_ref/feature.yaml:
```diff
@@ -1,8 +1,32 @@
+  kind: boolean
+  config_get: "show running | i feature"
+  config_get_token: '/^feature bgp$/'
+  config_set: "<state> feature bgp"
```
`<state>` isn't necessary since we don't need to disable features

chrisvanheuveln replied:
hey that was my comment on your PR! :-)

I'll fix.

robert-w-gries commented on lib/cisco_node_utils/cmd_ref/feature.yaml:
```diff
@@ -1,8 +1,32 @@
+    cli_nexus:
+      config_get: 'show running section feature'
+      config_get_token: '/^feature vn-segment-vlan-based$/'
+      config_set: 'feature vn-segment-vlan-based'
```
Why is `cli_nexus` needed here?

chrisvanheuveln replied:
I think it was a pre-emptive strike to help out the XR merge. But the `/N(3|9)/:` already keeps this cli out of XR. I'll remove it.

chrisvanheuveln committed:
Remove unneeded <state> refs

chrisvanheuveln committed:
Merge branch 'mv-rd-rt-from-bgp-to-vrf' of https://github.com/chrisvanheuveln/cisco-network-node-utils into mv-rd-rt-from-bgp-to-vrf

robert-w-gries commented:
:+1: but an extra pair of eyes wouldn't hurt
